### PR TITLE
feat: add rollback recovery for control plane resize

### PR DESCRIPTION
**Pull request by tuxerrante**
### Which issue this PR addresses:

Fixes [ARO-27429](https://redhat.atlassian.net/browse/ARO-27429)
Split from [#4805](https://github.com/Azure/ARO-RP/pull/4805) (3/3)

### What this PR does / why we need it:

Adds automatic rollback to `/resizecontrolplane` so failed control-plane resizes restore the cluster to a healthy state with 3 running, schedulable masters at matching SKUs.

Current logic in this PR:
- Pre-flight validation checks API server, etcd, SP, SKU/quota, and live control-plane inventory (Machine, Azure VM, and Node consistency) before the first mutation.
- Per-node snapshot now captures only the rollback state we need immediately before mutation: original VM size, Machine size, and schedulability.
- Forward path resizes nodes in reverse order (`master-2 -> master-1 -> master-0`) and gates each step on control-plane + etcd health.
- The linear `resizeNode()` happy path now reuses `pkg/util/steps` via `steps.Action(...)` / `steps.Run(...)`, while rollback state, retry policy, and operator-facing journaling stay local in `pkg/frontend`.
- `steps.Run` synthetic auth/Azure error rewrites are unwrapped locally so this admin path preserves its existing error contract instead of surfacing misleading synthetic `CloudError` responses.
- The etcd wait intentionally remains behind the local helper rather than `steps.Condition(...)`, so each health check keeps the existing per-check timeout boundary.
- Forward and rollback both rewrite the stable and beta node instance-type labels through the same helper, keeping label repair idempotent.
- Rollback restores node state in reverse order with Azure retries and step journaling.
- If etcd becomes unhealthy between rollback nodes, rollback progression now stops immediately so SRE can choose a targeted follow-up admin action.
- Retry behavior now uses the single production retry path; tests cover the same default-delay semantics we run at runtime.
- Admin reply preserves wrapped CloudError status/code/target while including rollback step context in message text.
- Focused unit coverage now pins both the synthetic-error handling around `steps.Run` and the `waitForEtcdHealthy()` timeout contract.

### Scope and reviewability notes

This branch was rebased on `master` and keeps the rollback feature scoped, while reusing the existing live inventory validation path instead of duplicating those checks inside snapshot capture.

### Suggested follow-up (separate PR)

To further reduce maintenance/review overhead without changing runtime behavior:
- Continue consolidating overlapping rollback test matrix cases in `resize_controlplane_test.go` vs `resize_controlplane_recovery_test.go`.

### Test plan for issue:

- `make fmt`
- `make lint-go`
- `make unit-test-go`
- `make test-go`

### Is there any documentation that needs to be updated for this PR?

No.

### How do you know this will function as expected in production?

Unit coverage validates pre-flight inventory gating, snapshot capture, forward/rollback sequencing, label restore behavior, error normalization, retry behavior, and etcd gating behavior for both forward and rollback paths.

### CODE REVIEW

**Comment by tuxerrante** [comment]
Thanks for the feedback — update after rebasing and cleanup pass:

- PR is now based on `master` and force-updated.
- Copilot feedback was applied with minimal behavior changes:
  - fail-fast on non-retryable etcd health errors,
  - deterministic/faster retry tests via injectable retry policy + `synctest`,
  - flaky deadline assertion fixed with `deadline.Sub(start)` tolerance.
- Kept Go 1.25 `range`-over-int form intentionally (locally verified).
- Reduced review noise by removing pre-validation carryover and duplicated rollback matrix tests from this PR scope.
- Added explicit rollback gate behavior: stop rollback progression if etcd becomes unhealthy between rollback nodes, so SRE can choose the next admin recovery action.

If this direction looks good, I’ll keep this as the final scope for #4865 and push any follow-up suggestions into a separate PR.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+	defer cancel()
+
+	return wait.PollImmediateUntilWithContext(ctx, etcdHealthPollInterval, func(ctx context.Context) (bool, error) {
+		if err := validateEtcdHealth(ctx, k); err != nil {
+			log.Infof("Waiting for etcd to become healthy: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+}
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:331]: Thanks for the callout. Fixed in `b88e4f2`: `waitForEtcdHealthy` now retries only for retryable etcd unhealthy (`409`) responses and returns non-retryable errors immediately, so we keep root-cause diagnostics instead of timing out with a generic poll error.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		t.Fatal("execution context should have a deadline")
+	}
+
+	expected := time.Duration(api.ControlPlaneNodeCount) * resizeControlPlanePerNodeTimeout
+	remaining := time.Until(deadline)
+	if remaining < expected-time.Second || remaining > expected {
+		t.Fatalf("execution deadline remaining %s, want close to %s", remaining, expected)
+	}
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go:429]: Agreed, this was flaky. Updated in `b88e4f2`: the test now captures `start` before context creation and asserts `deadline.Sub(start)` with a small tolerance window, avoiding scheduler/CI timing skew from `time.Until(deadline)`.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+		if attempt == azureOperationMaxAttempts-1 {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(azureOperationRetryDelay):
+		}
+	}
+	return fmt.Errorf("could not complete %s after %d attempts: %w", operationDesc, azureOperationMaxAttempts, lastErr)
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:358]: Good point. Addressed in `b88e4f2`: introduced a small internal retry policy so production keeps the same retry delay while tests can run with immediate retries. `TestRetryAzureOperation` now uses `synctest`, removing real sleep delays from CI.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+func ensureControlPlaneAndEtcdHealthy(ctx context.Context, k adminactions.KubeActions, nodeNames []string) error {
+	if err := ensureControlPlaneNodesReadyAndSchedulable(ctx, k, nodeNames); err != nil {
+		return err
+	}
+	return validateEtcdHealth(ctx, k)
+}
+
+func retryAzureOperation(ctx context.Context, operationDesc string, fn func() error) error {
+	var lastErr error
+	for attempt := range azureOperationMaxAttempts {
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:344]: Thanks. Kept `range` intentionally because this branch now targets Go 1.25+ where ranging over integers is valid. I verified locally with Go 1.25.3 as part of this update before rebasing/pushing.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane.go [hunk 5]
```diff
@@ -369,8 +367,16 @@ func doUpdateMachineVMSize(ctx context.Context, k adminactions.KubeActions, mach
 }
 
 func updateNodeInstanceTypeLabels(ctx context.Context, k adminactions.KubeActions, nodeName, vmSize string) error {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane.go]: Why do we need four functions to update the instancetype labels on a node? `updateNodeInstanceTypeLabels`, `setNodeInstanceTypeLabels` and `restoreNodeInstanceTypeLabels` are all very redundant.

I would suggest to only have `doSetNodeInstanceTypeLabels` and remove the ability to specify different values for the normal and beta node label.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+func (o *resizeControlPlaneOperation) captureNodeSnapshot(ctx context.Context, machineName string, machine machineValidationData) (controlPlaneNodeSnapshot, error) {
+	if machine.phase != "Running" {
+		return controlPlaneNodeSnapshot{}, api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"",
+			fmt.Sprintf("control plane machine %s is not Running (phase=%s)", machineName, machine.phase),
+		)
+	}
+	if machine.labelInstanceType == "" || machine.labelInstanceType != machine.size {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: The resize endpoint should be resilient enought to align the labels itself and not fail when they're set wrong to be begin with.

If we encounter this in the wild we always want to align the labels, so it's just annoying to have the resize admin endpoint tell us to do it manually.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+func (o *resizeControlPlaneOperation) recordStep(node, step string, d time.Duration, err error) {
+	if err != nil {
+		o.steps = append(o.steps, fmt.Sprintf("%s:%s failed (%s): %v", node, step, d.Truncate(time.Millisecond), err))
+	} else {
+		o.steps = append(o.steps, fmt.Sprintf("%s:%s (%s)", node, step, d.Truncate(time.Millisecond)))
+	}
+}
+
+func (o *resizeControlPlaneOperation) captureNodeSnapshot(ctx context.Context, machineName string, machine machineValidationData) (controlPlaneNodeSnapshot, error) {
+	if machine.phase != "Running" {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Shouldn't this be already covered by the pre-validation?

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+	vm, err := o.a.GetVirtualMachine(ctx, o.clusterResourceGroupName, machineName, mgmtcompute.InstanceView)
+	if err != nil {
+		return controlPlaneNodeSnapshot{}, fmt.Errorf("failed to capture Azure VM state for %s: %w", machineName, err)
+	}
+	if vm.VirtualMachineProperties == nil || vm.HardwareProfile == nil {
+		return controlPlaneNodeSnapshot{}, fmt.Errorf("failed to capture Azure VM state for %s: HardwareProfile missing", machineName)
+	}
+
+	actualVMSize := string(vm.HardwareProfile.VMSize)
+	if !strings.EqualFold(actualVMSize, machine.size) {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: We already ran the pre-validation where this is covered. There shouldn't be a need to check it again.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+
+	var node corev1.Node
+	if err := json.Unmarshal(rawNode, &node); err != nil {
+		return controlPlaneNodeSnapshot{}, fmt.Errorf("failed to parse node state for %s: %w", machineName, err)
+	}
+
+	labels := node.GetLabels()
+	nodeInstanceType := labels[nodeLabelInstanceType]
+	betaInstanceType := labels[nodeLabelBetaInstanceType]
+	if nodeInstanceType != betaInstanceType {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Why is this a problem? Let's just ignore it when the labels are different and make sure they're aligned afterwards

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+}
+
+func ensureControlPlaneAndEtcdHealthy(ctx context.Context, k adminactions.KubeActions, nodeNames []string) error {
+	if err := ensureControlPlaneNodesReadyAndSchedulable(ctx, k, nodeNames); err != nil {
+		return err
+	}
+	return validateEtcdHealth(ctx, k)
+}
+
+func retryAzureOperation(ctx context.Context, operationDesc string, fn func() error) error {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:342]: `retryAzureOperationWithPolicy` is only ever called with the default parameters `defaultRetryAzureOperationPolicy`, so there isn't really a reason to have a custom type and wrapping function for it. Especially since all of these functions and types aren't public and not intended to be used elsewhere, there isn't a need to make this more generic.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+}
+
+func (o *resizeControlPlaneOperation) rollbackNode(ctx context.Context, state *controlPlaneNodeProgress) error {
+	nodeName := state.snapshot.machineName
+	var rollbackErrs []error
+	// If resize didn't happen, there's nothing to restore.
+	vmSizeRestored := !state.vmResized
+	nodeReadyForSchedRestore := !state.vmStopped && !state.vmResized
+
+	if state.vmResized {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:370]: I don't believe that manually tracking the state of the resize and then checking which steps have been taken is a good pattern.
Instead, what i would suggest is that we design this with idempotency in mind. We just define the desired state of the machine via the `controlPlaneNodeSnapshot` type, and have this function just perform the operations neccary to align the current state with the desired state.

This way, we can even re-use the same state-align function for performing the resize and rolling back the changes, since they are essentially the same functionality.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+		start := time.Now()
+		err := func() error {
+			// Rollback always deallocates (true) regardless of the forward path's
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:373]: Why do we need to always de-allocate during rollback? This seems surprising to me, if i explicitly tell the resize admin action to not deallocate instances, it should respect that setting during the rollback as well.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+		"targetVMSize":   o.desiredVMSize,
+	}).Info("captured node snapshot for resize operation")
+
+	return snapshot, nil
+}
+
+func (o *resizeControlPlaneOperation) resizeNode(ctx context.Context, state *controlPlaneNodeProgress) error {
+	nodeName := state.snapshot.machineName
+
+	run := func(step string, fn func() error) error {
```
Comment (mrWinston) [inline pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Instead of rolling your own implementation, i would've liked to see the existing step functionality being reused and possibly extended to fit our use-case:
https://github.com/Azure/ARO-RP/tree/master/pkg/util/steps

Since it's already used for MiMo and the cluster installation, i don't think it makes a lot of sense to roll our own thing just for this admin action.

**Review by mrWinston — CHANGES_REQUESTED** [review]
Thanks for slimming it down, this is already much easier to review.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+		if policy.retryDelay <= 0 {
+			continue
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(policy.retryDelay):
+		}
+	}
+	return fmt.Errorf("could not complete %s after %d attempts: %w", operationDesc, policy.maxAttempts, lastErr)
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Won't fix for this PR scope: in production we always call `retryAzureOperation` with `defaultRetryAzureOperationPolicy`, where `retryDelay` is the non-zero constant `azureOperationRetryDelay` (5s). The zero-delay policy path is test-only injection, so this cancellation edge case is unlikely in runtime behavior and outside the rollback ticket scope.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane.go [hunk 15]
```diff
@@ -47,10 +49,33 @@ func (f *frontend) postAdminResizeControlPlane(w http.ResponseWriter, r *http.Re
...
+		return &api.CloudError{
+			StatusCode: cloudErr.StatusCode,
+			CloudErrorBody: &api.CloudErrorBody{
+				Code:    cloudErr.Code,
+				Message: err.Error(),
+				Target:  cloudErr.Target,
+				Details: cloudErr.Details,
+			},
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane.go:85]: Fixed in `50d210738`: `normalizeResizeControlPlaneErrorForAdminReply` now guards `cloudErr.CloudErrorBody == nil` and falls back safely (`InternalServerError` code with preserved status/message), so wrapped `*api.CloudError{StatusCode: ...}` no longer panics.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane.go (hunk 5 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane.go]: I took this in a few small follow-ups.

`5b3db6b0b` simplifies the helper surface so update and rollback both go through the same retrying setter, and `c75947b71` removes the empty-size fallback so the helper now fails fast instead of silently deleting labels.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 6 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: I want to double-check the intent here before changing that behavior: do we want `/resizecontrolplane` to intentionally ignore an already corrupted starting state and realign labels as part of the operation?

My understanding from the earlier discussions on these edge cases was to fail closed when the starting control-plane state is already unhealthy or inconsistent, so I kept that behavior here.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 10 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:342]: I cleaned this up in `394c035b0`.

`retryAzureOperation` is back to the single production path and the policy-only wrapper/test are gone, so the tests now cover the same retry behavior we use at runtime.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 6 above)
Comment (mrWinston) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: I just saw that we're indeed also checking this in the pre-validation and failing if the labels aren't aligned.
Like i said, i don't think this would be the best way to handle misaligned labels but since it's already on master, it's also not really in the scope of the PR.

However, since this is in fact already checked, i would still argue to remove the check here since we're just duplicating the logic from the pre-validation.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 7 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Yep, agreed. I moved this back into pre-flight in [27f269376](https://github.com/Azure/ARO-RP/commit/27f269376) by reusing the existing live Machine/Azure VM/Node inventory validator, so snapshot capture no longer re-checks it.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 8 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Yep, same cleanup in [27f269376](https://github.com/Azure/ARO-RP/commit/27f269376). The Machine/Azure VM consistency check now happens in pre-flight again, and `captureNodeSnapshot()` just records rollback state right before mutation.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 6 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: Thanks, I took this in [27f269376](https://github.com/Azure/ARO-RP/commit/27f269376).

The pre-flight path now reuses the live inventory validator again, and the duplicate snapshot-time label/size checks are gone. I kept the label setter itself idempotent so forward and rollback still just rewrite both labels from the target size.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 11 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:370]: I didn’t rewrite this into a fully generic state-align loop in this PR, but [27f269376](https://github.com/Azure/ARO-RP/commit/27f269376) does take a smaller step in that direction: pre-flight owns the live inventory validation again, snapshot capture now only records rollback state, and the node label updates stay idempotent.

I’d still rather keep a bigger control-flow rewrite or `pkg/util/steps` reuse separate from the rollback ticket.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 12 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go:373]: I dug a bit more here, and I think the reason to keep forced deallocation in the size-restore branch of rollback is that rollback has a different goal than the forward happy path.

[Azure’s VM resize docs](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/resize-vm) say that some size changes only work after deallocation, because the target size might not be available on the hardware cluster currently hosting the VM. 

They also say a failed resize can leave the VM model showing the requested size while the VM is still actually running on the previous size until allocation succeeds. That makes rollback a different problem than the forward happy path: once we are already in a failed or intermediate state, `deallocateVM=false` stops being a harmless preference and becomes an optimization that can reduce the chance of getting back to the known-good SKU. Relevant docs:

[VM restarting or resizing issues in Azure](https://learn.microsoft.com/en-us/troubleshoot/azure/virtual-machines/windows/restart-resize-error-troubleshooting)

In my opinion the Jira trail supports the same direction:

- [ARO-24547](https://redhat.atlassian.net/browse/ARO-24547): the rollback story’s DoD is to restore a healthy control plane with matching SKUs and 3 running masters. That points to “maximize recoverability” as the rollback goal.
- [ARO-22473](https://redhat.atlassian.net/browse/ARO-22473): explicitly says resize may need to be rerun from an intermediate state, and calls out that when a resize across SKU versions fails, SREs rerun it with the Deallocate VM option.

So my reasoning here is not that rollback should always ignore operator intent in general, but that for the VM-size-restore step it should take the most conservative Azure path to maximize recovery success. If we want, I can also tighten the inline comment to make that recoverability-first rationale clearer.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go (hunk 13 above)
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go]: I took the minimal-diff reuse path here.

The linear happy path in [`resizeNode()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L201-L314) now runs through [`runStep()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L139-L148) using `steps.Action(...)` / `steps.Run(...)`, while rollback-specific pieces such as [`captureNodeSnapshot()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L164-L199), [`retryAzureOperation()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L342-L359), [`rollbackNode()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L363-L480), and [`rollbackAll()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L482-L497) stay local.

One subtle issue showed up during the refactor: `steps.Run` rewrites certain auth/Azure errors before [`runStep()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L139-L148) ever sees them. In this admin resize flow that would have changed the client contract by surfacing misleading synthetic `CloudError` responses, so I kept the `steps` reuse but unwrap that runner-specific rewrite locally in [`unwrapSyntheticStepRunnerError()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L150-L160) to preserve the existing error semantics. The regression coverage is in [`TestRunStepPreservesOriginalActionErrors()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go#L103-L121).

I also intentionally did not retain `steps.Condition` for the etcd wait. `steps.Condition` evaluates its condition with the outer operation context, which would weaken the existing per-check timeout boundary owned by [`waitForEtcdHealthy()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L316-L331). To preserve current behavior, the etcd wait still runs through the local helper behind a `steps.Action(...)` inside [`resizeNode()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L201-L314), and I added [`TestWaitForEtcdHealthyBoundsHealthCheckByEtcdTimeout()`](https://github.com/Azure/ARO-RP/blob/8ace9225a971e3bf5292a6a13b784e617da13b9e/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go#L123-L159) to pin that timeout contract.

Validation run locally: `make lint-go`, `make unit-test-go`, and `make test-go`.

**Comment by tuxerrante** [comment]
Summary of today's updates on this PR:

- Pushed four reviewable commits:
  - [`6e187a482`](https://github.com/Azure/ARO-RP/commit/6e187a4820ff8045bc9a16b22a44015e7358ad81) `refactor(frontend): reuse steps in resize happy path`
  - [`8ace9225a`](https://github.com/Azure/ARO-RP/commit/8ace9225a971e3bf5292a6a13b784e617da13b9e) `test(frontend): pin etcd wait timeout in resize recovery`
  - [`af75bba00`](https://github.com/Azure/ARO-RP/commit/af75bba00) `fix(frontend): normalize resize admin error messages`
  - [`84a68e578`](https://github.com/Azure/ARO-RP/commit/84a68e578) `test(frontend): parallelize new resize regression tests`

- Reworked the linear forward path in [`resizeNode()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L201-L314) to reuse `pkg/util/steps` through [`runStep()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L139-L148), while keeping rollback-specific/stateful pieces local in [`captureNodeSnapshot()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L164-L199), [`retryAzureOperation()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L342-L359), [`rollbackNode()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L363-L480), and [`rollbackAll()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L482-L497).

- Kept the local guard around `steps.Run` synthetic auth/Azure rewrites in [`unwrapSyntheticStepRunnerError()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L150-L160), with coverage in [`TestRunStepPreservesOriginalActionErrors()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go#L103-L121), so the admin action preserves its existing error contract instead of surfacing misleading synthetic `CloudError` responses.

- Kept the etcd wait on the existing local helper instead of switching it to `steps.Condition(...)`, because this flow needs to preserve the per-check timeout boundary owned by [`waitForEtcdHealthy()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery.go#L316-L331). Added focused coverage in [`TestWaitForEtcdHealthyBoundsHealthCheckByEtcdTimeout()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go#L123-L159).

- Addressed the two remaining Copilot conversations with minimal scope:
  - [`normalizeResizeControlPlaneErrorForAdminReply()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane.go#L57-L108) now uses the wrapped `CloudErrorBody.Message` as the base admin-facing text when present and appends resize step / rollback context separately, so the response no longer duplicates status/code/target inside `message`. Regression coverage is in [`TestAdminReplyPreservesWrappedCloudError()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_resize_controlplane_recovery_test.go#L560-L615).
  - [`validateResizeControlPlaneInventory()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_vmresize_pre_validation.go#L178-L198) now normalizes joined inventory failures through [`convertErrorLineEndings()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_validatefullresize.go#L22-L30) before returning the `CloudError`, so Geneva/admin clients do not get embedded newlines. Coverage is in [`TestValidateResizeControlPlaneInventoryNormalizesJoinedErrorLineEndings()`](https://github.com/Azure/ARO-RP/blob/84a68e5782906d3a650fc15f6f08751455f060b2/pkg/frontend/admin_openshiftcluster_vmresize_pre_validation_test.go#L672-L730).

- Marked the two new regression tests parallel-safe in [`84a68e578`](https://github.com/Azure/ARO-RP/commit/84a68e578) to keep the extra coverage isolated to a test-only follow-up without changing production behavior.

- Updated the PR description and resolved the remaining Copilot review threads in-thread after pushing the final fixes.

Validation run locally after the final pushed state: `make fmt`, `make lint-go`, `make unit-test-go`, and `make test-go`.

File: pkg/frontend/admin_openshiftcluster_resize_controlplane.go [hunk 16]
```diff
@@ -47,10 +49,42 @@ func (f *frontend) postAdminResizeControlPlane(w http.ResponseWriter, r *http.Re
...
+		return &api.CloudError{
+			StatusCode: cloudErr.StatusCode,
+			CloudErrorBody: &api.CloudErrorBody{
+				Code:    code,
+				Message: err.Error(),
+				Target:  target,
+				Details: details,
+			},
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_resize_controlplane.go:85]: Took this in `af75bba00`.

`normalizeResizeControlPlaneErrorForAdminReply()` now uses the wrapped `CloudErrorBody.Message` as the base text when present and appends resize step / rollback context separately, so the admin response no longer duplicates status/code/target inside `message`.

The regression assertion is in `TestAdminReplyPreservesWrappedCloudError()`. Validation rerun on the branch after the follow-up test-only commit: `make fmt`, `make lint-go`, `make unit-test-go`, and `make test-go`.

File: pkg/frontend/admin_openshiftcluster_vmresize_pre_validation.go [hunk 17]
```diff
@@ -152,19 +153,50 @@ func (f *frontend) preResizeControlPlaneVMsValidation(
...
+) error {
+	err := validateLiveControlPlaneInventory(log, ctx, k, a, clusterResourceGroupID)
+	if err != nil {
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"controlPlaneInventory",
+			err.Error(),
+		)
+	}
```
[the remark this answers is not in the document]
Comment (tuxerrante) [reply pkg/frontend/admin_openshiftcluster_vmresize_pre_validation.go:222]: Took this in `af75bba00`.

`validateResizeControlPlaneInventory()` now runs the reused inventory validation error through `convertErrorLineEndings()` before returning the `CloudError`, so joined inventory failures stay single-line for Geneva/admin clients instead of carrying embedded newlines.

Coverage is in `TestValidateResizeControlPlaneInventoryNormalizesJoinedErrorLineEndings()`. `84a68e578` is only a small test-only follow-up marking the two new regression tests parallel-safe. Validation rerun on the branch: `make fmt`, `make lint-go`, `make unit-test-go`, and `make test-go`.

**Comment by tuxerrante** [comment]
**Rebased** this branch onto current `master` and force-pushed the updated history.

What changed:
- replayed the resize rollback series on top of `master`
- resolved the frontend resize/preflight conflicts against the newer validation shape on `master`
- folded the two post-rebase verification fixes into the rebased history instead of leaving them as a standalone follow-up commit
- final branch tip is `a7e048aa3` (`fix(frontend): preserve rebased resize preflight validation`)

Local validation after the rebase and fold-in:
- `make lint-go`
- `make unit-test-go`
- `make test-go`

All of the above passed locally before the force-push.

**Review by rh-returners — APPROVED** [review]
lgtm

**Review by mrWinston — APPROVED** [review]
Thanks for your changes! Looks good to me now.
